src/storage/src/source/mod.rs

```diff
@@ -931,12 +947,410 @@ where
+        waker: Waker,
+        persist_client: mz_persist_client::Client,
+        timestamp_frequency: Duration,
+        source_id: SourceInstanceId,
+        shard_id: ShardId,
+        bindings_channel: Sender<(PartitionId, MzOffset)>,
+    ) -> anyhow::Result<Option<Self>> {
+        // Untransmitted messages arranged by part.
+        // Messages are retained until their associated timestamp binding is durable,
+        // at which point they (and other messages at their time) flow out.
+        let messages_by_part = BTreeMap::new();
+
+        let partition_cursors = HashMap::new();
+        let read_cursors = HashMap::new();
+
+        let (mut write_handle, read_handle) = persist_client
+            .open(Duration::from_secs(10), shard_id)
+            .await
+            .expect("persist handles open err");
+
+        // Collection is closed.  No source to run
+        if read_handle.since().is_empty() || write_handle.upper().is_empty() {
+            return Ok(None);
+        }
+
+        let read_since = read_handle.since().clone();
+
+        // If this is the initialization of the persist collection, allow a read at Timestamp::minimum().  Else fetch
+        // the current upper.
+        let min_plus_one = Antichain::from_elem(Timestamp::minimum() + 1);
+        let write_upper = write_handle
+            .compare_and_append(
+                Duration::from_secs(10),
+                [],
+                Antichain::from_elem(Timestamp::minimum()),
+                min_plus_one.clone(),
+            )
+            .await
+            .expect("Initial CAA")
+            .with_context(|| "Initial compare and append")?
+            .err()
+            .unwrap_or(min_plus_one);
+
+        assert!(
+            timely::PartialOrder::less_equal(&read_since, &write_upper),
+            "{:?} PARTIAL ORDER: {:?} {:?}",
+            name,
+            read_since,
+            write_upper,
+        );
+
+        let mut snap_iter = read_handle
+            .snapshot(Duration::from_secs(10), read_since.clone())
+            .await
+            .expect(&format!("{:?} snapshot first", name))
+            .with_context(|| "Initial snapshot read")?;
+
+        let mut persisted_timestamp_bindings = BTreeMap::new();
+        loop {
+            let r = snap_iter.next(Duration::from_secs(2)).await;
+            match r {
+                Ok(v) if v.is_empty() => break,
+                Ok(v) => {
+                    for ((_key, value), timestamp, diff) in v {
+                        let partition = value.expect("Unable to decode partition id");
+                        persisted_timestamp_bindings
+                            .entry(partition)
+                            .or_insert(VecDeque::new())
+                            .push_back((timestamp, diff));
+                    }
+                }
+                Err(_) => break,
+            };
+        }
+        // TODO: bug (?) in persist that provides updates out of order?
```

Comment:
We don't give guarantees on the order of updates. The only thing to be relied on is that a progress message for `t` says there won't be updates with a `t'` smaller than `t`.

@danhhz for confirmation.

Reply:
Dan and I have a slack thread discussing this: https://materializeinc.slack.com/archives/C011P87EL2V/p1651608139281349

Reply:
this thread make me realize that it was easy for persist users to accidentally rely on the timestamps of a snapshot being unforwarded, but this will break once we have compaction. I threw together a PR that makes it more obvious, but I think it will break what this does: https://github.com/MaterializeInc/materialize/pull/12263

we should chat about how this should work (the summary is if you care about the timestamp an update was originally written at, you need to select an as_of for the snap+listener that's < the timestamp of the update). I understand that this was already working with stash, but I haven't yet gone and grokked exactly how yet

Reply:
maybe the three of us should hop on a call and hash it out?